--- C:/Soft/bblean/src/bbLean_mod_test_070426/src/blackbox/Tray.cpp	Sun Jun 12 22:02:00 2005
+++ C:/Soft/bblean/src/my/blackbox/Tray.cpp	Sat May 26 09:31:14 2007
@@ -537,16 +537,7 @@
     return FALSE;
 }
 
-//===========================================================================
-// Function: LoadShellServiceObjects
-// Purpose: Initializes COM, then starts all ShellServiceObjects listed in
-//          the ShellServiceObjectDelayLoad registry key, e.g. Volume,
-//          Power Management, PCMCIA, etc.
-// In: void
-// Out: void
-//===========================================================================
-
-ST void LoadShellServiceObjects()
+ST void WINAPI LoadShellServiceObjects(void* dummy)
 {
     // redundant by 'OleInitialize' in the main startup
     //CoInitializeEx(NULL, COINIT_APARTMENTTHREADED | COINIT_DISABLE_OLE1DDE);
@@ -597,6 +588,23 @@
 }
 
 //===========================================================================
+// Function: LoadShellServiceObjects
+// Purpose: Initializes COM, then starts all ShellServiceObjects listed in
+//          the ShellServiceObjectDelayLoad registry key, e.g. Volume,
+//          Power Management, PCMCIA, etc.
+// In: void
+// Out: void
+//===========================================================================
+
+// CyberShadow 2007.05.25: moved into a thread, since OLE initialization can take a significant amount of time on some systems, and causes the shell to freeze for up to a few minutes on start-up.
+ST void LoadShellServiceObjectsAsync()
+{
+    DWORD tid;
+    CloseHandle(CreateThread(NULL, 0, (LPTHREAD_START_ROUTINE)LoadShellServiceObjects, NULL, 0, &tid));
+}
+
+
+//===========================================================================
 // Function: UnloadShellServiceObjects
 //===========================================================================
 
@@ -692,7 +700,7 @@
 #ifdef BBOPT_DEVEL
         #pragma message("\n"__FILE__ "(665) : warning X0: LoadShellServiceObjects() disabled.\n")
 #else
-        LoadShellServiceObjects();
+        LoadShellServiceObjectsAsync();
 #endif
     }
 }
